### ISSUE TITLE
Unicode Support

### ISSUE THREAD

**Issue by Piping** [phillbush/xmenu#6]
Hi,

Currently I try to print some non ASCII character, and it shows box.

Is it a font choice or the program does not handle unicode display yet?

Thanks for this amazing tool.

**Comment by phillbush** [comment]
Xmenu does support Unicode display, so it is probably a font choice.

**Comment by Piping** [comment]
@phillbush if I have a font that I am known to have support for unicodes, how can I change the `config.h` to use this font?

**Comment by phillbush** [comment]
You can get a list of supported fonts with the command

    fc-list

It list the path to the font followed by a colon and a space and the font name. It is this font name that you should place between the quotation marks in `config.h`. Each font name can have some specifications appended with a colon, such as `:size=9`. For example, if I want the Helvetica font with size 10, I would write the following on `.config.h`:

    	.font = "Helvetica:size=11",

You can also use X resources instead of editing `config.h` and having to recompile xmenu.

Just put the following line to your `~/.Xresources` file:

    xmenu.font: Helvetica:size=11

For more information on the `~/.Xresources` file, see the [Arch Linux Wiki page on X Resources](https://wiki.archlinux.org/index.php/X_resources)

**Comment by Piping** [comment]
@phillbush Thanks for the detailed response.

I think I got the underlying cause. Does xmenu only support using *one font* right now?
There are always some characters missing from my menu if I try different font.

**Comment by phillbush** [comment]
Yes, xmenu does not support fallback fonts right now.  
I'm studying how to add support for more than one fonts for the next release.

**Comment by Piping** [comment]
@phillbush Great! There is one example I think of that dealt with the issue. st terminal source code `x.c`

**Comment by phillbush** [comment]
@Piping I am indeed using Suckless code (but dmenu's) for reference on getting multiple fonts with xft.
It is not an easy task, since I'm not using those "fancy libraries" such as pango, just plain xft+xlib.

**Comment by phillbush** [comment]
@Piping I've just added support for multiple fonts on xmenu with the next release.  
I needed to change a lot of code, which means it may appear new bugs, can you test if xmenu works for you and your fonts?

**Comment by Piping** [comment]
@phillbush  yeah sure. How can I specify the multi-font setup?

**Comment by phillbush** [comment]
Just separate the fonts with a comma, like

    	.font = "monospace:size=9,DejaVuSansMono:size=9",

You can also set the following in your `.Xresources` file (so you don't need to recompile it):

    xmenu.font: monospace:size=9,DejaVuSansMono:size=9

**Comment by Piping** [comment]
@phillbush very cool. 

It works for the following setup:
`xmenu.font: Ubuntu Mono:style=Regular,Noto Sans Mono CJK SC:style=Regular`
![image](https://user-images.githubusercontent.com/12042284/88865513-f56f7280-d1bc-11ea-8d25-ff7f507516eb.png)


It crashed for the setup:
`xmenu.font: Ubuntu Mono:style=Regular,Noto Sans Mono CJK SC:style=Regular,Noto Color Emoji:style=Regular`

Tested String: `Wed 29 Jul 2020 04:58:36 PM PDT 哈哈 🤣 `

**Comment by phillbush** [comment]
How does it crashes?
Does it prints some error message? Or does it just refuses to draw the menu?
I'm going to install those fonts here to test them.

**Comment by Piping** [comment]
@phillbush The menu did not draw. You can check out the font from https://www.nerdfonts.com/font-downloads.

**Comment by phillbush** [comment]
Does the crash only occurs when you are trying to render color emojis?  
If yes, this seems to be an error with xft.

If you use Arch Linux, try to install [libxft-bgra](https://aur.archlinux.org/packages/libxft-bgra/) (or the equivalent for your distro) and then recompile xmenu and see if the error persists.

**Comment by Piping** [comment]
@phillbush indeed. It seems like colored-emoji specific issue. I probably need to compile the library from source. It is strange that other application display color emoji fine. 

Is it possible to adjust the line height of the some font? These characters are not aligned horizontally.

**Comment by phillbush** [comment]
I fixed how xmenu computes the y position to draw the text in the newest commit, I think CJK characters are aligned horizontally now for most fonts.

**Comment by Piping** [comment]
@phillbush super!!

```
texty = y + (h - 4*(currfont->ascent + currfont->descent)/3)/2 + currfont->ascent + 5;
```
I tried this formula and it looks more aligned in my case. Recommend it!

I tried st with the colored emoji above and it crashed! But alacrity terminal works.

**Comment by phillbush** [comment]
> I tried st with the colored emoji above and it crashed! But alacrity terminal works.

So it is definitely a problem with libxft. There's [this guy](https://invidio.us/watch?v=f9qNXV01yzg) on youtube who made a video on this particular error.

**Comment by OliverLew** [comment]
> I fixed how xmenu computes the y position to draw the text in the newest commit, I think CJK characters are aligned horizontally now for most fonts.

Sorry if this sounds like nitpicking. I also ran into the same alignment problem of CJK characters in dwm then I spent some time digging into this. How I changed it finally is like this (in case of xmenu code):

```c
texty = y + (h + (dc.fonts[0]->ascent + dc.fonts[0]->descent))/2 - dc.fonts[0]->descent;
```

Most programs (I am not sure what they actually do) seem to be aligning the *baseline* of the glyphs. The complication is that the ascent and descent of CJK glyphs are usually bigger than Latin ones under the same fontsize (in my case they are de=19,ac=5 for Noto Sans CJK SC:size=12 and de=16,ac=5 for JetBrains Mono:size=12), making center align != baseline align. So this code set the baseline of all the glyphs the same as the first one when it is centered, it might not be perfect tho.

before: ![xmenu-old](https://user-images.githubusercontent.com/12032219/88997468-62504d00-d322-11ea-9422-10926a0875d5.png), after: ![xmenu-new](https://user-images.githubusercontent.com/12032219/88997484-6aa88800-d322-11ea-9828-404653a59b64.png)


Here is an image in [a blog](http://hayataki-masaharu.jp/web-typography-in-japanese/) showing Japanese typography:

![](https://hayataki-masaharu.jp/wp-content/uploads/2015/12/height2.png)

**Comment by phillbush** [comment]
>    texty = y + (h + (dc.fonts[0]->ascent + dc.fonts[0]->descent))/2 - dc.fonts[0]->descent;

@OliverLew, the problem with this formula of `texty` is that it depends solely on the first font, so if the user specify a CJK font before a Latin font, the problem will persist, and it will break the alignment if fonts with different geometries are used in sequence. The calculation of the height of each glyph must be based on the variable of the current font `currfont`.

I'm going to look for a better formula for computing `texty` that try to fix this problem. But I think it is not possible, for the glyph geometry depends on the font and whether the font sets a descender for CJK glyphs.

>    texty = y + (h - 4*(currfont->ascent + currfont->descent)/3)/2 + currfont->ascent + 5;

@Piping, the problem with this formula is that it depends on the geometry of the fonts you are using, in my font configuration, using your formula, the text does not render centralized on the menu item.

For now, I may add an option `-f` that makes the font alignment depends solely on the first font, as in @OliverLew formula, but this solution may be broken in the cases I specified.

I think you should send a mail to the suckless dev mailing list informing them about the problem on alignment of CJK characters and how they could solve it. The people there know more about libfontconfig, libXft (and fonts in general and their issues) than me.

If you want to change the character alignment for dwm, the `texty` variable is called `ty` there and belongs to the file `drw.c`, in the function `drw_text()` (just grep for a call to the function `XftDrawStringUtf8()`, which is on the line below it):

>					ty = y + (h - usedfont->h) / 2 + usedfont->xfont->ascent;

In dwm, `usedfont->h` is a variable for ascent + descent.

**Comment by OliverLew** [comment]
@phillbush Thank you for reply. Indeed I don't think there is a simple and elegant solution for this and I don't even recommend using my solution in your project.

> > texty = y + (h + (dc.fonts[0]->ascent + dc.fonts[0]->descent))/2 - dc.fonts[0]->descent;
> 
> the problem with this formula of `texty` is that it depends solely on the first font

Yes I totally agree and that is the main reason I don't think it is perfect.

> so if the user specify a CJK font before a Latin font, the problem will persist, and it will break the alignment if fonts with different geometries are used in sequence. The calculation of the height of each glyph must be based on the variable of the current font `currfont`.

My sole intention was to align the baselines by giving the `XftDrawStringUtf8` function a fixed `y` position to draw the characters instead of one depending on the different size of the glyphs. After that I fully rely on baseline align being the right way to position *all* kinds of characters, which I am not sure if it really is. As you suggested, I need to ask people who know more about this, or go researching for more existing implementations/standards.

And thanks for providing the help on dwm, actually I have already done that and have been using it for a long time maybe I delivered it in a misleading way :) but still I appreciate it very much, it could also help others who need it.

**Comment by Piping** [comment]
@phillbush Yeah. It a complicated issue for a generic solution. How about allow an option to set the formula used? Or allow an offset to adjust menu position for different geometry. (like the `+5` at the end of my formula.

**Comment by phillbush** [comment]
@OliverLew, I added the -f option, but I will remove it after I came with a better solution.

@Piping, adding an option for setting the formula would be cumbersome for parsing the terms in the formula, and adding an option for an offset only make sense if also adjusting the multipliers of the ascent and descent (and it would be hard to explain in the manual what each option does exactly). I think the easiest solution is the Suckless way: editing the source code for finer adjustments.
